internal/yunikorn/sync.go

```diff
@@ -84,16 +84,49 @@ func (s *Service) upsertPartitions(ctx context.Context) ([]*dao.PartitionInfo, e
 	}
 
 	err = s.workqueue.Add(func(ctx context.Context) error {
-		logger.Infow("upserting partitions", "count", len(partitions))
-		return s.repo.UpsertPartitions(ctx, partitions)
-	}, workqueue.WithJobName("upsert_partitions"))
+		logger.Infow("syncing partitions", "count", len(partitions))
+		err := s.repo.UpsertPartitions(ctx, partitions)
+		if err != nil {
+			return fmt.Errorf("could not upsert partitions: %w", err)
+		}
+		// Delete partitions that are not present in the API response
+		deleteCandidates, err := s.findPartitionDeleteCandidates(ctx, partitions)
```

Comment:
since workqueue has retries, you should probably build up `deleteCandidates` before pushing the task on it. If we ever retry, there is no need to re-calculate them